The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15736430336/job/44350330044) [(plain)](https://github.com/rust-lang/rust/commit/23e5243d84df3df394666af07e4e273754791dcb/checks/44350330044/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tidy check
`eslint` doesn't seem to be installed. Skipping tidy check for JS files.
You can install it using `npm install eslint@8.6.0`
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/src/ci/docker/host-x86_64/mingw-check-1/validate-rustdoc-json-format-version-update.sh:17: tab character
##[error]tidy error: /checkout/src/ci/docker/host-x86_64/mingw-check-1/validate-rustdoc-json-format-version-update.sh:18: tab character
##[error]tidy error: /checkout/src/ci/docker/host-x86_64/mingw-check-1/validate-rustdoc-json-format-version-update.sh:19: tab character
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:00
  local time: Wed Jun 18 15:12:22 UTC 2025
  network time: Wed, 18 Jun 2025 15:12:22 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/src/build_helper/src/git.rs
+++ b/src/build_helper/src/git.rs
@@ -198,7 +198,7 @@ fn get_latest_upstream_commit_that_modified_files(
 /// author.
 ///
 /// If we are in CI, we simply return our first parent.
-fn get_closest_upstream_commit(
+pub fn get_closest_upstream_commit(
     git_dir: Option<&Path>,
     config: &GitConfig<'_>,
     env: CiEnv,

--- a/src/tools/tidy/src/lib.rs
+++ b/src/tools/tidy/src/lib.rs
@@ -83,6 +83,7 @@ pub mod pal;
 pub mod rustdoc_css_themes;
 pub mod rustdoc_gui_tests;
 pub mod rustdoc_js;
+pub mod rustdoc_json;
 pub mod rustdoc_templates;
 pub mod style;
 pub mod target_policy;

--- a/src/tools/tidy/src/main.rs
+++ b/src/tools/tidy/src/main.rs
@@ -110,6 +110,7 @@ fn main() {
         check!(rustdoc_css_themes, &librustdoc_path);
         check!(rustdoc_templates, &librustdoc_path);
         check!(rustdoc_js, &librustdoc_path, &tools_path, &src_path);
+        check!(rustdoc_json);
         check!(known_bug, &crashes_path);
         check!(unknown_revision, &tests_path);
 

--- a/src/tools/tidy/src/main.rs
+++ b/src/tools/tidy/src/main.rs
@@ -110,7 +110,7 @@ fn main() {
         check!(rustdoc_css_themes, &librustdoc_path);
         check!(rustdoc_templates, &librustdoc_path);
         check!(rustdoc_js, &librustdoc_path, &tools_path, &src_path);
-        check!(rustdoc_json);
+        check!(rustdoc_json, &src_path);
         check!(known_bug, &crashes_path);
         check!(unknown_revision, &tests_path);
 

--- a/src/tools/tidy/src/rustdoc_json.rs
+++ b/src/tools/tidy/src/rustdoc_json.rs
@@ -3,16 +3,39 @@
 
 use std::process::Command;
 
+use build_helper::ci::CiEnv;
+use build_helper::git::{GitConfig, get_closest_upstream_commit};
+use build_helper::stage0_parser::parse_stage0_file;
+
 fn git_diff(base_commit: &str, extra_arg: &str) -> Option<String> {
     let output = Command::new("git").arg("diff").arg(base_commit).arg(extra_arg).output().ok()?;
     Some(String::from_utf8_lossy(&output.stdout).into())
 }
 
 pub fn check(bad: &mut bool) {
-    let Ok(base_commit) = std::env::var("BASE_COMMIT") else {
-        // Not in CI so nothing we can check here.
-        println!("not checking rustdoc JSON `FORMAT_VERSION` update");
-        return;
+    println!("Checking tidy rustdoc_json...");
+    let stage0 = parse_stage0_file();
+    let base_commit = match get_closest_upstream_commit(
+        None,
+        &GitConfig {
+            nightly_branch: &stage0.config.nightly_branch,
+            git_merge_commit_email: &stage0.config.git_merge_commit_email,
+        },
+        CiEnv::current(),
+    ) {
+        Ok(Some(commit)) => commit,
+        Ok(None) => {
+            *bad = true;
+            eprintln!("No base commit found, skipping rustdoc_json check");
+            return;
+        }
+        Err(error) => {
+            *bad = true;
+            eprintln!(
+                "Failed to retrieve base commit for rustdoc_json check because of `{error}`, skipping it"
+            );
+            return;
+        }
     };
 
     // First we check that `src/rustdoc-json-types` was modified.

--- a/src/tools/tidy/src/rustdoc_json.rs
+++ b/src/tools/tidy/src/rustdoc_json.rs
@@ -1,0 +1,66 @@
+//! Tidy check to ensure that `FORMAT_VERSION` was correctly updated if `rustdoc-json-types` was
+//! updated as well.
+
+use std::process::Command;
+
+fn git_diff(base_commit: &str, extra_arg: &str) -> Option<String> {
+    let output = Command::new("git").arg("diff").arg(base_commit).arg(extra_arg).output().ok()?;
+    Some(String::from_utf8_lossy(&output.stdout).into())
+}
+
+pub fn check(bad: &mut bool) {
+    let Ok(base_commit) = std::env::var("BASE_COMMIT") else {
+        // Not in CI so nothing we can check here.
+        println!("not checking rustdoc JSON `FORMAT_VERSION` update");
+        return;
+    };
+
+    // First we check that `src/rustdoc-json-types` was modified.
+    match git_diff(&base_commit, "--name-status") {
+        Some(output) => {
+            if !output
+                .lines()
+                .any(|line| line.starts_with("M") && line.contains("src/rustdoc-json-types"))
+            {
+                // `rustdoc-json-types` was not modified so nothing more to check here.
+                return;
+            }
+        }
+        None => {
+            *bad = true;
+            eprintln!("Failed to run `git diff`");
+            return;
+        }
+    }
+    // Then we check that if `FORMAT_VERSION` was updated, the `Latest feature:` was also updated.
+    match git_diff(&base_commit, "src/rustdoc-json-types") {
+        Some(output) => {
+            let mut format_version_updated = false;
+            let mut latest_feature_comment_updated = false;
+            for line in output.lines() {
+                if line.starts_with("+pub const FORMAT_VERSION: u32 =") {
+                    format_version_updated = true;
+                } else if line.starts_with("+// Latest feature:") {
+                    latest_feature_comment_updated = true;
+                }
+            }
+            if format_version_updated != latest_feature_comment_updated {
+                *bad = true;
+                if latest_feature_comment_updated {
+                    eprintln!(
+                        "`Latest feature` comment was updated whereas `FORMAT_VERSION` wasn't"
+                    );
+                } else {
+                    eprintln!(
+                        "`Latest feature` comment was not updated whereas `FORMAT_VERSION` was"
+                    );
+                }
+            }
+        }
+        None => {
+            *bad = true;
+            eprintln!("Failed to run `git diff`");
+            return;
+        }
+    }
+}

--- a/src/tools/tidy/src/rustdoc_json.rs
+++ b/src/tools/tidy/src/rustdoc_json.rs
@@ -1,18 +1,20 @@
 //! Tidy check to ensure that `FORMAT_VERSION` was correctly updated if `rustdoc-json-types` was
 //! updated as well.
 
+use std::ffi::OsStr;
+use std::path::Path;
 use std::process::Command;
 
 use build_helper::ci::CiEnv;
 use build_helper::git::{GitConfig, get_closest_upstream_commit};
 use build_helper::stage0_parser::parse_stage0_file;
 
-fn git_diff(base_commit: &str, extra_arg: &str) -> Option<String> {
+fn git_diff<S: AsRef<OsStr>>(base_commit: &str, extra_arg: S) -> Option<String> {
     let output = Command::new("git").arg("diff").arg(base_commit).arg(extra_arg).output().ok()?;
     Some(String::from_utf8_lossy(&output.stdout).into())
 }
 
-pub fn check(bad: &mut bool) {
+pub fn check(src_path: &Path, bad: &mut bool) {
     println!("Checking tidy rustdoc_json...");
     let stage0 = parse_stage0_file();
     let base_commit = match get_closest_upstream_commit(
@@ -26,13 +28,13 @@ pub fn check(bad: &mut bool) {
         Ok(Some(commit)) => commit,
         Ok(None) => {
             *bad = true;
-            eprintln!("No base commit found, skipping rustdoc_json check");
+            eprintln!("error: no base commit found for rustdoc_json check");
             return;
         }
         Err(error) => {
             *bad = true;
             eprintln!(
-                "Failed to retrieve base commit for rustdoc_json check because of `{error}`, skipping it"
+                "error: failed to retrieve base commit for rustdoc_json check because of `{error}`"
             );
             return;
         }
@@ -46,17 +48,18 @@ pub fn check(bad: &mut bool) {
                 .any(|line| line.starts_with("M") && line.contains("src/rustdoc-json-types"))
             {
                 // `rustdoc-json-types` was not modified so nothing more to check here.
+                println!("`rustdoc-json-types` was not modified.");
                 return;
             }
         }
         None => {
             *bad = true;
-            eprintln!("Failed to run `git diff`");
+            eprintln!("error: failed to run `git diff` in rustdoc_json check");
             return;
         }
     }
     // Then we check that if `FORMAT_VERSION` was updated, the `Latest feature:` was also updated.
-    match git_diff(&base_commit, "src/rustdoc-json-types") {
+    match git_diff(&base_commit, src_path.join("rustdoc-json-types")) {
         Some(output) => {
             let mut format_version_updated = false;
             let mut latest_feature_comment_updated = false;
@@ -71,19 +74,18 @@ pub fn check(bad: &mut bool) {
                 *bad = true;
                 if latest_feature_comment_updated {
                     eprintln!(
-                        "`Latest feature` comment was updated whereas `FORMAT_VERSION` wasn't"
+                        "error: `Latest feature` comment was updated whereas `FORMAT_VERSION` wasn't"
                     );
                 } else {
                     eprintln!(
-                        "`Latest feature` comment was not updated whereas `FORMAT_VERSION` was"
+                        "error: `Latest feature` comment was not updated whereas `FORMAT_VERSION` was"
                     );
                 }
             }
         }
         None => {
             *bad = true;
-            eprintln!("Failed to run `git diff`");
-            return;
+            eprintln!("error: failed to run `git diff` in rustdoc_json check");
         }
     }
 }

